### PR TITLE
Add geometry functions

### PR DESCRIPTION
**Pull request by livace**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add geometry functions for work with polygons, serialization and deserialization

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

### CODE REVIEW

File: src/Functions/geometryConverters.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        throw Exception("Failed to cast " + col.type->getName() + " to " + wanted_data_type->getName(), ErrorCodes::ILLEGAL_COLUMN);
+    }
+    return Parser(std::move(casted));
+}
+
+}
+
+GeometryFromColumnParser makeGeometryFromColumnParser(const ColumnWithTypeAndName & col)
+{
+    LOG_FATAL(&Poco::Logger::get("geoconv"), "Fine");
```
Comment (akuzm) [inline src/Functions/geometryConverters.cpp:69]: :)

File: src/Functions/geometryConverters.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+#include <common/logger_useful.h>
+
+namespace DB {
+
+namespace {
+
+size_t getArrayDepth(DataTypePtr data_type, size_t max_depth)
+{
+    LOG_FATAL(&Poco::Logger::get("geoconv"), "getting depth");
```
Comment (akuzm) [inline src/Functions/geometryConverters.cpp:12]: Debug logs here as well.

File: src/Functions/PolygonsUnion.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) override
+    {
+        auto get_parser = [&block, &arguments] (size_t i) {
+            const auto * const_col =
+                checkAndGetColumn<ColumnConst>(block.getByPosition(arguments[i]).column.get());
+
+            bool is_const = static_cast<bool>(const_col);
+
+            return std::pair<bool, GeometryFromColumnParser>{is_const, is_const ?
+                makeGeometryFromColumnParser(ColumnWithTypeAndName(const_col->getDataColumnPtr(), block.getByPosition(arguments[i]).type, block.getByPosition(arguments[i]).name)) :
```
Comment (akuzm) [inline src/Functions/PolygonsUnion.cpp:80]: Don't understand this part -- what is the argument column type, text? And you are parsing it into boost geometry?

File: src/Functions/geometryConverters.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        case 3: return makeParser<DataTypeCustomMultiPolygonSerialization, Float64MultiPolygonFromColumnParser>(col);
+        default: throw Exception("Cannot parse geometry from column with type " + col.type->getName()
+                + ", array depth is too big", ErrorCodes::ILLEGAL_COLUMN);
+    }
+}
+
+Float64Geometry createContainer(const GeometryFromColumnParser & parser)
+{
+    static ContainerCreator creator;
+    return boost::apply_visitor(creator, parser);
```
Comment (akuzm) [inline src/Functions/geometryConverters.cpp:83]: What is this line doing? It looks like just some very convoluted way to say `return parser.createContainer()`.

File: src/Functions/geometryConverters.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+void get(const GeometryFromColumnParser & parser, Float64Geometry & container, size_t i)
+{
+    boost::apply_visitor(Getter(container, i), parser);
```
Comment (akuzm) [inline src/Functions/geometryConverters.cpp:88]: Ditto the above comment.

**Comment by akuzm** [comment]
I like the branch name, but this is not a good size for a PR -- hard to review, and, judging by the amount of debug output left, even for yourself. If it has some smaller part that makes sense by itself (e.g. a couple of new functions + infractructure for them), it would be better to split it out and commit first.

We'll probably need some brief docs for these functions (just a sentence about each function is enough, for starters), and also a performance test.

By the way, I see that you are repacking geometry data from our format into boost format -- how bad is this performance-wise? Maybe we could somehow tweak our data structure to have identical memory layout?

**Comment by alexey-milovidov** [comment]
Geometry data types should be parametrized by "metric".
Metric can be: `Euclidean` and `LocalWGS84`.

Example: `Point('LocalWGS84')`.

When we do any calculation on geometry type, we should do the following:
- compute the bounding box of all geometries participating in the calculation (for all function arguments as a whole);
- compute its midpoint;
- obtain the coefficients of metric tensor at the midpoint;
- use it in calculation (e.g. when calculating the area, multiply it by both of the coefficients).

If metric tensor is trivial (for `Euclidean` metric), we can omit the calculation of midpoint.

**Comment by nikitamikhaylov** [comment]
> LocalWGS84

@alexey-milovidov Could you please explain a little more precise and detailed why do we need "do the following" on every calculation. Why `Euclidean` and `LocalWGS84` are metrics? How do the differ? I googled and watched some presentation, but I don't really understand how it correlates with mathematical definition of metric. 

I my view, we have several coordinate systems (2d plane, sphere or WGS84) And we have some policy how to calculate distance, for example. If we have two points (let them be in coordinate system with longitude and latitude), we can threat them as points on sphere or as points on ellipsoid. So, we can tune the accuracy of the result. And that is all we need.

**Comment by alexey-milovidov** [comment]
Metric in the sense of "metric tensor". It defines scalar product and allows to calculate lengths of vectors (and, consequently, lengths of curves). We can treat it as additional coefficients that should be applied when we calculate lengths. Example: https://en.wikipedia.org/wiki/Metric_tensor#The_round_metric_on_a_sphere

**Comment by alexey-milovidov** [comment]
> we have several coordinate systems (2d plane, sphere or WGS84) And we have some policy how to calculate distance, for example

That is the definition of metric, exactly.

**Comment by alexey-milovidov** [comment]
More details:

Suppose you want to calculate scalar product of two vectors.
You can treat first vector as row and second vector as column and do matrix multiplication:

`vec1_row * vec2_column`

You can also calculate the length of the vector: `sqrt(scalar_product(vec, vec))`
But in some coordinate systems you have to calculate scalar product differently to get correct length.
It can be represented as a matrix in between:

`vec1_row * metric_matrix * vec2_column`

`metric_matrix` is the metric tensor. It's also a function of location.
(if we want to calculate the length of a curve, we have to take lengths of small tangent vectors in every point and integrate them)

It is symmetrical matrix. Moreover, for all our practical cases, it is also diagonal.
And we have just two coefficients.

And for sphere and ellipsoid, the coefficient for longitude is always 1.
So, the only one coefficient remains.

**Comment by hamedtayeban** [comment]
12chzBKTn5pcj1GNvtVZriKgEQKEv2CKZe

**Pull request closed without merging**